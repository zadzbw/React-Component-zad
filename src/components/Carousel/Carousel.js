/**
 * Created by zad on 17/4/20.
 */
import React, {Component, PropTypes, Children} from 'react';
import classNames from 'classnames';
import _isFunction from 'lodash/isFunction';
import isOneOf from '../../utils/isOneOf';
import keyCode from '../../utils/keyCode';
import {rightShift} from '../../utils/array_shift';

const carouselClass = 'zad-carousel';

export default class Carousel extends Component {
  constructor(props) {
    super(props);
    this.state = {
      current: 0,
      prevCurrent: this.getChildrenLength() - 1,
    };
  }

  static propTypes = {
    showDots: PropTypes.bool,
    onPageChange: PropTypes.func,
  };

  static defaultProps = {
    showDots: true,
    onPageChange: () => undefined,
  };

  getChildrenLength = () => {
    return this.props.children.length;
  };

  // [0, 1, 2, -1]
  getPositionArray = () => {
    const result = [];
    const len = this.getChildrenLength();
    for (let i = 0; i < len - 1; i++) {
      result.push(i);
    }
    result.push(-1);
    return result;
  };

  setCurrent = (v) => {
    const {current} = this.state;
    const {onPageChange} = this.props;
    if (v !== current) {
      this.setState({
        current: v,
        prevCurrent: current,
      });
      if (_isFunction(onPageChange)) {
        onPageChange({
          current: v,
          prevCurrent: current,
        });
      }
    }
  };

  getNextCurrent = (flag = true) => {
    const {current} = this.state;
    const len = this.getChildrenLength();
    if (flag) {
      return (current + 1) % len;
    } else {
      return (current + len - 1) % len;
    }
  };

  _keyDown = (e) => {
    const code = e.keyCode;
    if (isOneOf([keyCode.LEFT, keyCode.UP], code)) {
      e.preventDefault();
      const prevCurrent = this.getNextCurrent(false);
      this.setCurrent(prevCurrent);
    } else if (isOneOf([keyCode.RIGHT, keyCode.DOWN], code)) {
      e.preventDefault();
      const nextCurrent = this.getNextCurrent(true);
      this.setCurrent(nextCurrent);
    }
  };

  getItems = () => {
    const {children} = this.props;
    const {current, prevCurrent} = this.state;
    const positionArray = rightShift(this.getPositionArray(), current);
    return Children.map(children, (child, i) => {
      const itemClass = classNames({
        [`${carouselClass}-item`]: true,
        [`${carouselClass}-item-active`]: current === i,
      });
      const position = positionArray[i];
      const min = Math.min(current, prevCurrent);
      const max = Math.max(current, prevCurrent);
      return React.cloneElement(child, {
        className: itemClass,
        style: {
          transform: `translate3d(${ position * 100}%, 0px, 0px)`,
          transitionDuration: (min <= i && i <= max) ? '400ms' : '0ms',
        }
      });
    });
  };

  getDots = () => {
    const {children} = this.props;
    const {current} = this.state;
    return Children.map(children, (child, i) => {
      const dotClass = classNames({
        [`${carouselClass}-dot`]: true,
        [`${carouselClass}-dot-active`]: current === i,
      });
      return (
        <li className={dotClass}>
          <button onClick={this.setCurrent.bind(this, i)}/>
        </li>
      );
    });
  };

  render() {
    const {showDots} = this.props;
    return (
      <div className={`${carouselClass}-wrapper`} tabIndex="0" onKeyDown={this._keyDown}>
        <div className={`${carouselClass}-content`}>
          {this.getItems()}
        </div>
        <div className={`${carouselClass}-nav`}>

        </div>
        {
          showDots &&
          <ul className={`${carouselClass}-dots`}>
            {this.getDots()}
          </ul>
        }
      </div>
    );
  }
}
