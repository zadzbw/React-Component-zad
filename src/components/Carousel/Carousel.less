@import "../../style/variables";
@import "../../style/mixins/index";

@carousel-class: zad-carousel;
@dots-height: 3px;

.@{carousel-class} {
  &-wrapper {
    position: relative;
    overflow: hidden;
    outline: none;
  }

  &-content {
    position: relative;
    height: 200px;
  }

  &-item {
    position: absolute;
    display: inline-block;
    top: 0;
    left: 0;
    width: 100%;
    height: 100%;
    transition: transform 0.4s @ease-in-out;
    z-index: 0;

    &-active {
      z-index: 3;
    }
  }

  &-nav {

  }

  &-dots {
    position: absolute;
    list-style: none;
    bottom: 12px;
    text-align: center;
    width: 100%;
    height: @dots-height;
    z-index: 5;
    .clearfix();

    .@{carousel-class}-dot {
      position: relative;
      display: inline-block;
      margin: 0 2px;
      padding: 0;
      vertical-align: top;

      button {
        display: block;
        border: 0;
        outline: none;
        cursor: pointer;
        width: 16px;
        height: @dots-height;
        border-radius: 1px;
        background-color: #fff;
        opacity: 0.4;
        transition: all 0.4s @ease-in-out;

        &:hover {
          opacity: 0.75;
        }
      }

      &-active {
        button {
          width: 24px;
          opacity: 1 !important;
        }
      }
    }
  }
}
