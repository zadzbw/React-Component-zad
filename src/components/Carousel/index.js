/**
 * Created by zad on 17/4/20.
 */
import './Carousel.less';
import Carousel from './Carousel';

export default Carousel;
