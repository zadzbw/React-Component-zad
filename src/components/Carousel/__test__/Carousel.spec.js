/**
 * Created by zad on 17/4/21.
 */
import React from 'react';
import {shallow} from 'enzyme';
import {shallowToJson} from 'enzyme-to-json';
import Carousel from '../Carousel';

describe('Carousel', () => {
  it('normal', () => {
    const wrapper = shallow(
      <Carousel>
        <div><h3>0</h3></div>
        <div><h3>1</h3></div>
        <div><h3>2</h3></div>
        <div><h3>3</h3></div>
        <div><h3>4</h3></div>
        <div><h3>5</h3></div>
      </Carousel>
    );

    expect(wrapper.state('current')).toEqual(0);
    expect(wrapper.find('.zad-carousel-dot').at(0).hasClass('zad-carousel-dot-active')).toBeTruthy();
    wrapper.find('.zad-carousel-dot').at(3).find('button').simulate('click'); // dot click
    expect(wrapper.state('current')).toEqual(3);
    expect(wrapper.find('.zad-carousel-dot').at(3).hasClass('zad-carousel-dot-active')).toBeTruthy();

    expect(shallowToJson(wrapper)).toMatchSnapshot();
  });

});