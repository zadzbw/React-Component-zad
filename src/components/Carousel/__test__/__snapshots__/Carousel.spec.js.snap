exports[`Carousel normal 1`] = `
<div
  className="zad-carousel-wrapper"
  onKeyDown={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  tabIndex="0">
  <div
    className="zad-carousel-content">
    <div
      className="zad-carousel-item"
      style={
        Object {
          "transform": "translate3d(300%, 0px, 0px)",
          "transitionDuration": "400ms",
        }
      }>
      <h3>
        0
      </h3>
    </div>
    <div
      className="zad-carousel-item"
      style={
        Object {
          "transform": "translate3d(400%, 0px, 0px)",
          "transitionDuration": "400ms",
        }
      }>
      <h3>
        1
      </h3>
    </div>
    <div
      className="zad-carousel-item"
      style={
        Object {
          "transform": "translate3d(-100%, 0px, 0px)",
          "transitionDuration": "400ms",
        }
      }>
      <h3>
        2
      </h3>
    </div>
    <div
      className="zad-carousel-item zad-carousel-item-active"
      style={
        Object {
          "transform": "translate3d(0%, 0px, 0px)",
          "transitionDuration": "400ms",
        }
      }>
      <h3>
        3
      </h3>
    </div>
    <div
      className="zad-carousel-item"
      style={
        Object {
          "transform": "translate3d(100%, 0px, 0px)",
          "transitionDuration": "0ms",
        }
      }>
      <h3>
        4
      </h3>
    </div>
    <div
      className="zad-carousel-item"
      style={
        Object {
          "transform": "translate3d(200%, 0px, 0px)",
          "transitionDuration": "0ms",
        }
      }>
      <h3>
        5
      </h3>
    </div>
  </div>
  <Animate
    animation={Object {}}
    component="span"
    componentProps={Object {}}
    onAppear={[Function]}
    onEnd={[Function]}
    onEnter={[Function]}
    onLeave={[Function]}
    transitionAppear={false}
    transitionEnter={true}
    transitionLeave={true}
    transitionName="move-left" />
  <Animate
    animation={Object {}}
    component="span"
    componentProps={Object {}}
    onAppear={[Function]}
    onEnd={[Function]}
    onEnter={[Function]}
    onLeave={[Function]}
    transitionAppear={false}
    transitionEnter={true}
    transitionLeave={true}
    transitionName="move-right" />
  <ul
    className="zad-carousel-dots">
    <li
      className="zad-carousel-dot">
      <button
        onClick={[Function]} />
    </li>
    <li
      className="zad-carousel-dot">
      <button
        onClick={[Function]} />
    </li>
    <li
      className="zad-carousel-dot">
      <button
        onClick={[Function]} />
    </li>
    <li
      className="zad-carousel-dot zad-carousel-dot-active">
      <button
        onClick={[Function]} />
    </li>
    <li
      className="zad-carousel-dot">
      <button
        onClick={[Function]} />
    </li>
    <li
      className="zad-carousel-dot">
      <button
        onClick={[Function]} />
    </li>
  </ul>
</div>
`;
